pkg/tsdb/prometheus/client/client.go

```diff
@@ -77,36 +78,43 @@ func (c *Client) QueryExemplars(ctx context.Context, q *models.Query) (*http.Res
 	qs.Set("start", formatTime(tr.Start))
 	qs.Set("end", formatTime(tr.End))
 
-	return c.fetch(ctx, c.method, u, qs)
+	return c.fetch(ctx, c.method, u, qs, nil)
 }
 
-func (c *Client) QueryResource(ctx context.Context, method string, p string, qs url.Values) (*http.Response, error) {
-	u, err := url.ParseRequestURI(c.baseUrl)
+type FetchReq struct {
+	Method      string
+	Url         *url.URL
+	QueryString url.Values
+}
+
+func (c *Client) QueryResource(ctx context.Context, req *backend.CallResourceRequest) (*http.Response, error) {
+	baseUrlParsed, err := url.ParseRequestURI(c.baseUrl)
 	if err != nil {
 		return nil, err
 	}
+	reqUrlParsed, err := url.Parse(req.URL)
 
-	u.Path = path.Join(u.Path, p)
+	baseUrlParsed.Path = path.Join(baseUrlParsed.Path, req.URL)
+	baseUrlParsed.RawQuery = reqUrlParsed.RawQuery
 
-	return c.fetch(ctx, method, u, qs)
+	return c.fetch(ctx, req.Method, baseUrlParsed, nil, req.Body)
 }
 
-func (c *Client) fetch(ctx context.Context, method string, u *url.URL, qs url.Values) (*http.Response, error) {
-	if strings.ToUpper(method) == http.MethodGet {
+func (c *Client) fetch(ctx context.Context, method string, u *url.URL, qs url.Values, body []byte) (*http.Response, error) {
+	if strings.ToUpper(method) == http.MethodGet && qs != nil {
 		u.RawQuery = qs.Encode()
 	}
-
-	r, err := http.NewRequestWithContext(ctx, method, u.String(), nil)
+	bodyReader := bytes.NewReader(body)
+	request, err := http.NewRequestWithContext(ctx, method, u.String(), bodyReader)
 	if err != nil {
 		return nil, err
 	}
 
 	if strings.ToUpper(method) == http.MethodPost {
-		r.Body = ioutil.NopCloser(strings.NewReader(qs.Encode()))
```

Comment:
the `fetch` method was originally meant to be used by the query API methods to handle the switching between GET or POST based on what the user selected in the data source settings, so removing this will break POST bodies for those.

this might be a bad idea, but i attempted to simplify this [on this branch](https://github.com/grafana/grafana/blob/0af51eee721488fce80750735e07215c6ffdd87b/pkg/tsdb/prometheus/client/client.go) by implementing the `http.RoundTripper` interface and building the full resource call proxied `http.Request` in `resource.go`, and calling `RoundTrip` directly instead of mixing the proxy code with prometheus API client methods

Reply:
> so removing this will break POST bodies for those methods. i think it got a bit confusing once resource calls started using fetch since they pull the HTTP method from a different source

This does not remove that, method is still read from datasource settings. From what I can tell this only makes sure we forward the request body in case of POST

Reply:
> so removing this will break POST bodies for those methods. i think it got a bit confusing once resource calls started using fetch since they pull the HTTP method from a different source

This does not remove that, method is still read from datasource settings. From what I can tell this only makes sure we forward the request body in case of POST

Reply:
The decision on what the method is should be on the front end for now IMHO. So this just forwards the request as it is removing as much magic in between as possible. We still strip some headers which I guess makes sense and make sure the URL is correct but otherwise, I don't see any need to alter the request when doing a resource call.

So I assume this was meant as a way to change GET resource request to POST in the grafana server but we don't really do that and I am not sure if that would give us many benefits.

Reply:
Also if you want POST based on config, what would be the benefit doing a GET between browser - grafana and then switch to POST between grafana - prometheus? Seems like that translation is not really necessary and could just lead to problems and confusion.

Reply:
@torkelo @aocenas ah ok, i think i might be misunderstanding how the POST setting is supposed to be used for `/api/ds/query` calls which use the QueryRange, QueryInstant, and QueryExemplars methods